### PR TITLE
update transaction overview

### PR DESCRIPTION
**Pull request by jackysp**
Signed-off-by: Shuaipeng Yu <jackysp@gmail.com>

### Which TiDB version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [x] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### CODE REVIEW

**Comment by TomShawn** [comment]
@jackysp Please specify the applicable versions, thanks!

**Comment by jackysp** [comment]
> @jackysp Please specify the applicable versions, thanks!

But I've checked with @yikeke , the pr for sw will be merged to the sw branch, and then merge all the modifications to master, 4.0, 3.1, 3.0 together. Maybe some big PRs.

**Comment by yikeke** [comment]
> > @jackysp Please specify the applicable versions, thanks!
> 
> But I've checked with @yikeke , the pr for sw will be merged to the sw branch, and then merge all the modifications to master, 4.0, 3.1, 3.0 together. Maybe some big PRs.

Sorry, I think I meant the docs-special-week PRs will be merged to master and 4.0 branches, not 3.1 and 3.0 branches. So if your PR changes apply to v3.1, v3.0, or v2.1, you need to tick the checkboxes in the PR description. @jackysp

File: transaction-overview.md [hunk 1]
```diff
@@ -37,6 +37,10 @@ START TRANSACTION WITH CONSISTENT SNAPSHOT;
 
 以上三条语句都用于开启事务，效果相同。执行开启事务语句可以显式地开启一个新的事务。如果执行以上语句时，当前 Session 正处于一个事务的中间过程，那么系统会先自动提交当前事务，再开启一个新的事务。
 
+> **注意：**
+>
+> 与 MySQL 不同的是，TiDB 在执行完上述语句后即会获取当前数据库快照，而 MySQL 的 `BEGIN` 和 `START TRANSACTION` 是在开启事务后的第一个从 InnoDB 读数据的 SELECT 语句（非 SELECT FOR UPDATE）后获取，`START TRANSACTION WITH CONSISTENT SNAPSHOT` 是语句执行时获取。因此，TiDB 中的 `BEGIN`、`START TRANSACTION` 和 `START TRANSACTION WITH CONSISTENT SNAPSHOT` 都等效为 MySQL 中的 `START TRANSACTION WITH CONSISTENT SNAPSHOT`。
```
Comment (TomShawn) [inline transaction-overview.md:42]:
```suggestion
> 与 MySQL 不同的是，TiDB 在执行完上述语句后即会获取当前数据库快照，而 MySQL 的 `BEGIN` 和 `START TRANSACTION` 是在开启事务后的第一个从 InnoDB 读数据的 `SELECT` 语句（非 `SELECT FOR UPDATE`）后获取快照，`START TRANSACTION WITH CONSISTENT SNAPSHOT` 是语句执行时获取快照。因此，TiDB 中的 `BEGIN`、`START TRANSACTION` 和 `START TRANSACTION WITH CONSISTENT SNAPSHOT` 都等效为 MySQL 中的 `START TRANSACTION WITH CONSISTENT SNAPSHOT`。
```

File: transaction-overview.md [hunk 2]
```diff
@@ -151,54 +156,17 @@ rollback;
...
-START TRANSACTION;
-UPDATE my_table SET a ='new_value' WHERE id = 1;
-UPDATE my_table SET a ='newer_value' WHERE id = 2;
-UPDATE my_table SET a ='newest_value' WHERE id = 3;
-COMMIT;
-```
+TiDB 默认设置了单个事务的容量的总大小不超过 100 MB，这个默认值可以通过配置文件中的配置项 `txn-total-size-limit` 进行修改，最大支持到 10 GB。实际的单个事务大小限制还取决于服务器剩余可用内存大小，执行事务时 TiDB 进程的内存消耗大约是事务大小 6 倍以上。
 
-同理，执行 `INSERT` 语句时，建议使用显式事务。
+在 4.0 以前的版本，TiDB 限制了单个事务的键值对的总数量不超过 30 万条，在 4.0 之后的版本没有这项限制。
```
Comment (TomShawn) [inline transaction-overview.md:167]:
```suggestion
在 4.0 以前的版本，TiDB 限制了单个事务的键值对的总数量不超过 30 万条，从 4.0 版本起 TiDB 取消了这项限制。
```

File: transaction-overview.md [hunk 3]
```diff
@@ -151,54 +156,18 @@ rollback;
...
-UPDATE my_table SET a ='newer_value' WHERE id = 2;
-UPDATE my_table SET a ='newest_value' WHERE id = 3;
-```
+由于底层存储引擎的限制，TiDB 要求单行不超过 6 MB。可以将一行的所有列根据类型转换为字节数并加和来估算单行大小。
 
-此时每一条语句都需要经过两阶段提交，频繁的网络交互致使延迟率高。为提升事务执行效率，可以选择使用显式事务，即在一个事务内执行三条语句。
+TiDB 同时支持乐观事务与悲观事务，其中乐观事务是悲观事务的基础。由于乐观事务是先将修改缓存在私有内存中，因此，TiDB 对于单个事务的容量做了限制。
 
-优化后版本：
+TiDB 默认设置了单个事务的容量的总大小不超过 100 MB，这个默认值可以通过配置文件中的配置项 `txn-total-size-limit` 进行修改，最大支持到 10 GB。实际的单个事务大小限制还取决于服务器剩余可用内存大小，执行事务时 TiDB 进程的内存消耗大约是事务大小 6 倍以上。
```
Comment (kissmydb) [inline transaction-overview.md:165]: 这里同时提一下如果开启 binlog ，txn-total-size-limit 的区别吧

File: transaction-overview.md [hunk 4]
```diff
@@ -151,54 +156,18 @@ rollback;
...
-* 客户端在提交之前，数据都写在内存中，而数据量过多时易导致 OOM (Out of Memory) 错误。
-* 在第一阶段写入数据耗时增加，与其他事务出现写冲突的概率会指数级增长。
-* 最终导致事务完成提交的耗时增加。
-
-为了使性能达到最优，建议每 100～500 行写入一个事务。
-
-TiDB 设置了单个事务的键值对的总大小不超过 100 MB，这个默认值可以通过配置文件中的配置项 `txn-total-size-limit` 进行修改，最大支持到 10GB。实际的单个事务大小限制还取决于用户的内存，执行大事务时 TiDB 进程的内存消耗大约是事务大小 6 倍以上。
-
-在 4.0 以前的版本，TiDB 限制了单个事务的键值对的总数量不超过 30 万条，在 4.0 之后的版本没有这项限制。
+> 以 Kafka 为例，Kafka 的单条消息处理能力的上限是 1 GB。因此，当把 `txn-total-size-limit` 设置为 1 GB 以上时，可能出现事务在 TiDB 中执行成功，但下游 Kafka 报错的情况。为避免这种情况出现，请用户根据最终消费者的限制来决定 `txn-total-size-limit` 的实际大小。例如：下游使用了 Kafka，则 `txn-total-size-limit` 不应超过 1 GB。
```
Comment (kissmydb) [inline transaction-overview.md:173]: 之前记得 1G 是 binlog 的硬限制，现在是说 binlog 同步到 tidb，10G 也是可以的对么？

Comment (jackysp) [reply transaction-overview.md:173]: 对。

**Comment by jackysp** [comment]
PTAL @TomShawn @kissmydb

File: transaction-overview.md [hunk 5]
```diff
@@ -120,12 +124,13 @@ SELECT * FROM T; -- MySQL 返回 1 2；TiDB 返回 1
 惰性检查的意义在于，如果对事务中每个 `INSERT` 语句都立刻进行唯一性约束检查，将造成很高的网络开销。而在提交时进行一次批量检查，将会大幅提升性能。
 
 > **注意：**
->
-> 本优化仅对普通的 `INSERT` 语句生效，对 `INSERT IGNORE` 和 `INSERT ON DUPLICATE KEY UPDATE` 不会生效。
+> 
+> 1. 本优化仅在乐观事务中生效。
```
Comment (TomShawn) [inline transaction-overview.md:128]:
```suggestion
> + 本优化仅在乐观事务中生效。
```

File: transaction-overview.md [hunk 6]
```diff
@@ -120,12 +124,13 @@ SELECT * FROM T; -- MySQL 返回 1 2；TiDB 返回 1
...
+> 2. 本优化仅对普通的 `INSERT` 语句生效，对 `INSERT IGNORE` 和 `INSERT ON DUPLICATE KEY UPDATE` 不会生效。
```
Comment (TomShawn) [inline transaction-overview.md:129]:
```suggestion
> + 本优化仅对普通的 `INSERT` 语句生效，对 `INSERT IGNORE` 和 `INSERT ON DUPLICATE KEY UPDATE` 不会生效。
```

**Pull request merged**